jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/sender/JMSClientConnector.java

```diff
@@ -77,76 +77,141 @@ public boolean send(CarbonMessage carbonMessage, CarbonCallback carbonCallback)
-                throw new ClientConnectorException(e.getMessage(), e);
-            }
-
-            Message message = null;
-            String messageType = propertyMap.get(JMSConstants.JMS_MESSAGE_TYPE);
 
-            if (carbonMessage instanceof TextCarbonMessage) {
-                String textData = ((TextCarbonMessage) carbonMessage).getText();
-                if (messageType.equals(JMSConstants.TEXT_MESSAGE_TYPE)) {
-                    message = session.createTextMessage();
-                    TextMessage textMessage = (TextMessage) message;
-                    textMessage.setText(textData);
-                } else if (messageType.equals(JMSConstants.BYTES_MESSAGE_TYPE)) {
-                    message = session.createBytesMessage();
-                    BytesMessage bytesMessage = (BytesMessage) message;
-                    bytesMessage.writeBytes(textData.getBytes(Charset.defaultCharset()));
-                }
-            } else if (messageType.equals(JMSConstants.OBJECT_MESSAGE_TYPE) &&
-                       carbonMessage instanceof SerializableCarbonMessage) {
-                message = session.createObjectMessage((SerializableCarbonMessage) carbonMessage);
-            } else if (messageType.equals(JMSConstants.MAP_MESSAGE_TYPE) && carbonMessage instanceof MapCarbonMessage) {
-                message = session.createMapMessage();
-                MapMessage mapMessage = (MapMessage) message;
-                Enumeration<String> mapNames = ((MapCarbonMessage) carbonMessage).getMapNames();
-                while (mapNames.hasMoreElements()) {
-                    String key = mapNames.nextElement();
-                    mapMessage.setString(key, ((MapCarbonMessage) carbonMessage).getValue(key));
-                }
-            }
+            setupMessageProducer(propertyMap.entrySet());
 
-            if (carbonMessage.getProperty(JMSConstants.PERSISTENCE) != null &&
-                carbonMessage.getProperty(JMSConstants.PERSISTENCE).equals(false)) {
-                messageProducer.setDeliveryMode(DeliveryMode.NON_PERSISTENT);
-            }
+            Message jmsMessage = createJmsMessage(carbonMessage, propertyMap);
 
-            if (message != null) {
-                messageProducer.send(message);
-            } else {
-              throw new ClientConnectorException("Exception occured while creating the message");
-            }
+            sendMessage(carbonMessage, jmsMessage);
 
-        } catch (JMSException e) {
-            throw new ClientConnectorException("Exception occurred while sending the message", e);
         } finally {
             if (jmsConnectionFactory != null) {
                 try {
                     jmsConnectionFactory.closeMessageProducer(messageProducer);
                     jmsConnectionFactory.closeSession(session);
                     jmsConnectionFactory.closeConnection(connection);
                 } catch (JMSConnectorException e) {
```

Comment:
Depending on the connection factory nature (caching, pooling, none) it will decide to close the session.